### PR TITLE
feat: add lib for dev count

### PR DESCRIPTION
**Pull request by maxjeffos**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
Upon `snyk monitor`, run git log in the direcotry `snyk monitor` is run; this means that that we are assuming it is run from a directory which is in a git repo. We git log only the last 90 days. Look at the hash(author email) and the commit timestamps. We create a map of (hashed author email) -> (latest commit timestamp for that author). Send that along with the API call to Snyk that is already used in monitor.

#### Any background context you want to provide?
This is for the "contributing developer" count feature.

#### What are the relevant tickets?
HAMMER-36

#### Related PRs
https://github.com/snyk/registry/pull/13742

### CODE REVIEW

File: src/lib/monitor/dev-count-analysis.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  const commitTimestamp = unwrapQuotedString(quotedCommitTimestamp);
+  const commitInfo = new GitCommitInfo(authorEmail, commitTimestamp);
+  return commitInfo;
+}
+
+export function parseGitLog(gitLog: string): GitRepoCommitStats {
+  if (gitLog.trim() === '') {
+    return GitRepoCommitStats.empty();
+  }
+  const logLines = gitLog.trim().split('\n'); // previously used os.EOL but that didn't work right on Windows
```
Comment (JackuB) [inline src/lib/monitor/dev-count-analysis.ts:100]: btw it's a configurable option to use different newline character https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration#_formatting_and_whitespace

So maybe you'll have to check both?

Comment (maxjeffos) [reply src/lib/monitor/dev-count-analysis.ts:100]: Ya, good call. I also need to test it on a Windows machine to see what I get by default with real data.

Comment (maxjeffos) [reply src/lib/monitor/dev-count-analysis.ts]: fixed

File: src/cli/commands/monitor/index.ts [hunk 2]
```diff
@@ -80,6 +87,29 @@ async function monitor(...args0: MethodArgs): Promise<any> {
...
+        CONTRIBUTING_DEVELOPER_PERIOD_DAYS,
+      );
+      const gitLogResults = await runGitLog(
+        timestampStartOfContributingDeveloperPeriod,
+        repoPath,
+      );
+      const stats: GitRepoCommitStats = parseGitLog(gitLogResults);
+      repoContributors = stats.getUniqueHashedEmailsAndMostRecentCommitTimestamps();
+    } catch (err) {
+      console.error(err);
```
Comment (JackuB) [inline src/cli/commands/monitor/index.ts:109]: I’d avoid outputting anything from here unless in debug. Console output (unrelated to monitor command they are trying to run) could confuse users.

I’d check if we have a way to post these errors to our logz.io

File: src/lib/monitor/dev-count-analysis.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
+import * as crypto from 'crypto';
```
Comment (JackuB) [inline src/lib/monitor/dev-count-analysis.ts:1]: I would add here a rather large comment to explain why we do it and what we collect.

File: src/lib/monitor/index.ts [hunk 4]
```diff
@@ -48,6 +48,9 @@ interface MonitorBody {
   target: {};
   targetFileRelativePath: string;
   targetFile: string;
+  repoContributors?: {
```
Comment (JackuB) [inline src/lib/monitor/index.ts:51]: Registry is expecting a different format - array of hash and last commit timestamp. See my PR on registry for key names

File: src/lib/monitor/dev-count-analysis.ts [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+  const lookbackTimespanSeconds = timespanInDays * ONE_DAY_IN_SECONDS;
+  const startOfPeriodEpochSeconds = nowUtcEpocS - lookbackTimespanSeconds;
+  return startOfPeriodEpochSeconds;
+}
+
+export async function runGitLog(
+  timestampEpochSecondsStartOfPeriod: number,
+  repoPath: string,
+): Promise<string> {
+  const gitLogCommand = `git --no-pager log --pretty=tformat:"'%H'${SERIOUS_DELIMITER}'%an'${SERIOUS_DELIMITER}'%ae'${SERIOUS_DELIMITER}'%aI'" --after="${timestampEpochSecondsStartOfPeriod}"`;
```
Comment (JackuB) [inline src/lib/monitor/dev-count-analysis.ts:167]: Why even have the quotes if you are separating by delimiter?

File: test/dev-count-analysis.jest-test.ts [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+  const line =
+    "'0bd4d3c394a54ba54f6c44705ac73d7d87b39525'_SNYK_SEPARATOR_'some-user'_SNYK_SEPARATOR_'someemail-1@somedomain.com'_SNYK_SEPARATOR_'2020-02-06T11:43:11+00:00'";
+  const commitInfo: GitCommitInfo = parseGitLogLine(line);
+  t.equal(
+    commitInfo.authorHashedEmail,
+    expectedEmailHashes['someemail-1@somedomain.com'],
+  );
+});
+
+test('can handle an empty git log', (t) => {
```
Comment (JackuB) [inline test/dev-count-analysis.jest-test.ts:77]: Would also add a case where git is not available and command throws (?)

File: src/lib/monitor/index.ts (hunk 4 above)
Comment (maxjeffos) [reply src/lib/monitor/index.ts]: fixed

File: src/lib/monitor/dev-count-analysis.ts (hunk 3 above)
Comment (maxjeffos) [reply src/lib/monitor/dev-count-analysis.ts:1]: I don't think you can even use the Snyk CLI unless you've authenticated it... so not sure the part about "authenticated users" is needed. What do you think?

File: src/lib/monitor/dev-count-analysis.ts (hunk 5 above)
Comment (maxjeffos) [reply src/lib/monitor/dev-count-analysis.ts]: took them out

File: test/dev-count-analysis.jest-test.ts (hunk 6 above)
Comment (maxjeffos) [reply test/dev-count-analysis.jest-test.ts:85]: done

File: src/cli/commands/monitor/index.ts (hunk 2 above)
Comment (maxjeffos) [reply src/cli/commands/monitor/index.ts]: Fixed the console.log; looking into the logzio part.

**Pull request merged**